recipe-server/client/control/components/ExtensionForm.js

```diff
@@ -0,0 +1,137 @@
+import { Map } from 'immutable';
+import React, { PropTypes as pt } from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import { initialize, reduxForm } from 'redux-form';
+
+import QueryExtension from './data/QueryExtension';
+import { ControlField } from './Fields';
+import { showNotification } from '../actions/NotificationActions';
+import { createExtension, updateExtension } from '../state/extensions/actions';
+import { getExtension } from '../state/extensions/selectors';
+
+
+class ExtensionForm extends React.Component {
+  static propTypes = {
+    createExtension: pt.func.isRequired,
+    extensionId: pt.number,
+    handleSubmit: pt.func.isRequired,
+    initializeForm: pt.func.isRequired,
+    initialValues: pt.object,
+    showNotification: pt.func.isRequired,
+    updateExtension: pt.func.isRequired,
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.handleSave = ::this.handleSave;
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { initialValues, initializeForm } = this.props;
+
+    if (initialValues.id !== nextProps.initialValues.id) {
+      initializeForm('extension', nextProps.initialValues);
+    }
+  }
+
+  async handleSave(values) {
+    const { extensionId } = this.props;
+    const data = values;
+
+    // Remove the XPI value passed through because this is the URL of the existing XPI
+    if (data.xpi) {
+      delete data.xpi;
+    }
+
+    // Store the newly selected XPI with the key `xpi`.
+    if (data.xpiUpload) {
+      data.xpi = data.xpiUpload[0];
+      delete data.xpiUpload;
+    }
+
+    try {
+      if (extensionId) {
+        await this.props.updateExtension(extensionId, data);
+      } else {
+        await this.props.createExtension(data);
+      }
+    } catch (error) {
+      this.props.showNotification({
+        messageType: 'error',
+        message: 'Extension cannot be saved. Please correct any errors listed in the form below.',
+      });
+    }
+
+    this.props.showNotification({
```

Comment:
Shouldn't this be inside the try so that it doesn't show when there's a failure?